# <p align="center">Siemens Sales Forecast</p>

---

## <p align="center">*1 - Exploratory Data Analysis & Pre-processing*</p>

---

### 👥 **Team Members**
- **Ana Farinha** *(Student Number: 20211514)*  
- **António Oliveira** *(Student Number: 20211595)*  
- **Mariana Neto** *(Student Number: 20211527)*  
- **Salvador Domingues** *(Student Number: 20240597)*  

📅 **Date:** *April 1, 2025*  
📍 **Prepared for:** *Siemens*  

**GitHub Repo:** https://github.com/MGN19/Siemens-forecast

---

# ToC

<a class="anchor" id="top"></a>


1. [Import Libraries & Data](#1.-Import-Libraries-&-Data) <br><br>

2. [Data Exploration](#2.-Data-Exploration) <br>

    2.1 [Descriptive Statistics](#2.1-Descriptive-Statistics) <br>
    
    2.2 [Duplicate Records](#2.2-Duplicate-Records) <br>
   
    2.3 [Missing Data Analysis](#2.3-Missing-Data-Analysis) <br><br>

3. [Feature Exploration](#3.-Feature-Exploration) <br>

    3.1 [Univariate Analysis](#3.1-Univariate-Analysis) <br>
    
    3.2 [Bivariate Analysis](#3.2-Bivariate-Analysis) <br><br>

4. [Data Cleaning & Preprocessing](#4.-Data-Cleaning-&-Preprocessing) <br>

    4.1 [Treating Incoherencies](#4.1-Treating-Incoherencies) <br>
    
    4.2 [Missing Values](#4.2-Missing-Values) <br>

    4.3 [Feature Engineering](#4.3-Feature-Engineering) <br>

    &emsp; 4.3.1 [Feature Creation](#4.3.1-Feature-Creation) <br>

    &emsp; 4.3.2 [Encoding](#4.3.2-Encoding) <br>

    4.4 [Outliers](#4.4-Outliers) <br>

    &emsp; 4.4.1 [Univariate Outliers](#4.4.1-Univariate-Outliers) <br>

    &emsp; 4.4.2 [Multivariate Outliers](#4.4.2-Multivariate-Outliers) <br>

    4.5 [Feature Selection](#4.5-Feature-Selection) <br>

    &emsp; 4.5.1 [Dropping Features](#4.5.1-Dropping-Features) <br>

    4.6 [Data Exploration](#4.6-Data-Exploration)<br>

    &emsp; 4.6.1 [Univariate Analysis](#4.6.1-Univariate-Analysis) <br>

    &emsp; 4.6.2 [Bivariate Analysis](#4.6.2-Bivariate-Analysis) <br> 

# 1. Import Libraries & Data

In [3]:
# Data Handling Libraries
import pandas as pd
import numpy as np
pd.set_option('display.max_columns', None)

# from ydata_profiling import ProfileReport

# Visualisation libraries
import missingno as msno
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

# Data Preprocessing Libraries
from sklearn.preprocessing import OneHotEncoder, RobustScaler
from sklearn.impute import KNNImputer

# Suppress Warnings
import warnings
warnings.filterwarnings("ignore")

# Custom Functions
#import functions as f
#import utils as u

**Data**

In [5]:
sales_data = pd.read_csv('./data/Case2_Sales data.csv', 
                 sep = ';')
sales_data

DATE Mapped_GCK Sales_EUR
0     01.10.2018         #1         0
1     02.10.2018         #1         0
2     03.10.2018         #1         0
3     04.10.2018         #1         0
4     05.10.2018         #1         0
...          ...        ...       ...
9797  23.08.2019        #12         0
9798  23.08.2019        #36      1015
9799  12.08.2019        #12         0
9800  28.08.2019         #8  43763,91
9801  27.08.2019         #8         0

[9802 rows x 3 columns]

In [14]:
test_data = pd.read_csv('data/Case2_Test Set Template.csv', 
                 sep = ';')
test_data

Month Year Mapped_GCK  Sales_EUR
0       Mai 22         #3        NaN
1       Jun 22         #3        NaN
2       Jul 22         #3        NaN
3       Aug 22         #3        NaN
4       Sep 22         #3        NaN
..         ...        ...        ...
135     Okt 22         #1        NaN
136     Nov 22         #1        NaN
137     Dez 22         #1        NaN
138     Jan 23         #1        NaN
139     Feb 23         #1        NaN

[140 rows x 3 columns]

In [11]:
market_data = pd.read_excel('./data/Case2_Market data.xlsx', index_col = 'Unnamed: 0')
market_data

China  \
Index 2010=100 (if not otherwise noted)  Production Index Machinery & Electricals   
date                                                               MAB_ELE_PRO156   
 2004m2                                                                 16.940704   
 2004m3                                                                 23.711852   
 2004m4                                                                 24.435235   
...                                                                           ...   
2021m12                                                                310.763183   
 2022m1                                                                235.956129   
 2022m2                                                                235.956129   
 2022m3                                                                329.413367   
 2022m4                                                                267.373145   

                                                                         China.1  \
Index 2010=100 (if not otherwise noted)  Shipments Index Machinery & Electricals   
date                                                              MAB_ELE_SHP156   
 2004m2                                                                16.940704   
 2004m3                                                                23.711852   
 2004m4                                                                24.435235   
...                                                                          ...   
2021m12                                                               310.763183   
 2022m1                                                               235.956129   
 2022m2                                                               235.956129   
 2022m3                                                               329.413367   
 2022m4                                                               267.373145   

                                                                           France  \
Index 2010=100 (if not otherwise noted)  Production Index Machinery & Electricals   
date                                                               MAB_ELE_PRO250   
 2004m2                                                                112.091273   
 2004m3                                                                136.327976   
 2004m4                                                                117.791806   
...                                                                           ...   
2021m12                                                                100.565744   
 2022m1                                                                 85.743503   
 2022m2                                                                  90.60354   
 2022m3                                                                107.843548   
 2022m4                                                                  87.69811   

                                                                        France.1  \
Index 2010=100 (if not otherwise noted)  Shipments Index Machinery & Electricals   
date                                                              MAB_ELE_SHP250   
 2004m2                                                                83.458866   
 2004m3                                                               106.168192   
 2004m4                                                                92.007646   
...                                                                          ...   
2021m12                                                               134.589504   
 2022m1                                                                108.15632   
 2022m2                                                                117.71577   
 2022m3                                                                136.85872   
 2022m4                                                               116.528738   

                

#### **Production Index (Machinery & Electricals)**
| Country        | Code              |
|---------------|------------------|
| China        | MAB_ELE_PRO156    |
| France       | MAB_ELE_PRO250    |
| Germany      | MAB_ELE_PRO276    |
| Italy        | MAB_ELE_PRO380    |
| Japan        | MAB_ELE_PRO392    |
| Switzerland  | MAB_ELE_PRO756    |
| United Kingdom | MAB_ELE_PRO826  |
| United States | MAB_ELE_PRO840   |
| Europe       | MAB_ELE_PRO1100   |

#### **Shipments Index (Machinery & Electricals)**
| Country        | Code              |
|---------------|------------------|
| China        | MAB_ELE_SHP156    |
| France       | MAB_ELE_SHP250    |
| Germany      | MAB_ELE_SHP276    |
| Italy        | MAB_ELE_SHP380    |
| Japan        | MAB_ELE_SHP392    |
| Switzerland  | MAB_ELE_SHP756    |
| United Kingdom | MAB_ELE_SHP826  |
| United States | MAB_ELE_SHP840   |
| Europe       | MAB_ELE_SHP1100   |

#### **Material Prices**
| Indicator                         | Code                  |
|-----------------------------------|----------------------|
| Price of Base Metals             | RohiBASEMET1000_org  |
| Price of Energy                  | RohiENERGY1000_org   |
| Price of Metals & Minerals       | RohiMETMIN1000_org   |
| Price of Natural Gas Index       | RohiNATGAS1000_org   |
| Price of Crude Oil, Average      | RohCRUDE_PETRO1000_org |
| Price of Copper                  | RohCOPPER1000_org    |

#### **Producer Prices / Production code: Electrical Equipment**
| Country         | Producer price Code         | Production Code         |
|-----------------|------------------|------------------|
| United States   | PRI27840_org     | PRO27840_org     |
| United Kingdom  | PRI27826_org     | PRO27826_org     |
| Italy           | PRI27380_org     | PRO27380_org     |
| France          | PRI27250_org     | PRO27250_org     |
| Germany         | PRI27276_org     | PRO27276_org     |
| China           | PRI27156_org     | -                |
| World           | -                | PRO271000_org    |
| Switzerland     | -                | PRO27756_org     |
| Japan           | -                | PRO27392_org     |


#### **Production code: Machinery and Equipment**
| Country        | Code              |
|---------------|------------------|
| United States | PRO28840_org     |
| World        | PRO281000_org     |
| Switzerland  | PRO28756_org      |
| United Kingdom | PRO28826_org    |
| Italy        | PRO28380_org      |
| Japan        | PRO28392_org      |
| France       | PRO28250_org      |
| Germany      | PRO28276_org      |

#### **United States: EUR in LCU**
| Country        | Code              |
|---------------|------------------|
| United States | WKLWEUR840_org   |


# 2. Data Exploration

<a href="#top">Top &#129033;</a>

**Profiling Report**

In [ ]:
# profile = ProfileReport(
#     df, 
#     title='Data',
#     correlations={
#         "pearson": {"calculate": True},
#         "spearman": {"calculate": False},
#         "kendall": {"calculate": False},
#         "phi_k": {"calculate": False},
#         "cramers": {"calculate": False},
#     },
# )

# profile

## 2.1 Descriptive Statistics

<a href="#top">Top &#129033;</a>

<br>

**2.1.1 Exploring the dataset structure with .info() method**

The `.info()` method provides more information about the dataset, including the existence of missing values and the data type of the elements of each row.

In [23]:
sales_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9802 entries, 0 to 9801
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   DATE        9802 non-null   object
 1   Mapped_GCK  9802 non-null   object
 2   Sales_EUR   9802 non-null   object
dtypes: object(3)
memory usage: 229.9+ KB


With the information above, we can see the following:
- all of the columns in this dataset are categorical variables. For analysis purposes, the `DATE` column will be transformed into a date format, and the `Sales_EUR` column into a float. 
- we can also see that there might not be any missing values in any of the columns of the dataset.

Based on this, we will first change the data types of these 2 columns identified so that analysis can be proprerly made.

In [25]:
# Converting the DATE column to a datetime format
sales_data['DATE'] = pd.to_datetime(sales_data['DATE'], format='%d.%m.%Y')

# Converting the Sales_EUR to a numeric format
sales_data['Sales_EUR'] = pd.to_numeric(sales_data['Sales_EUR'].str.replace(',', '.'))


<br>

2.1.2 Statistics

By using the `describe()` method we get different statistical measurements, such as mean and standard deviation, for the various features. We will include the parameter 'all' so both statistics for numerical and categorical data appear. 

In [29]:
sales_data.describe(include='all').T

count unique  top  freq                           mean  \
DATE          9802    NaN  NaN   NaN  2020-07-19 17:55:48.785961984   
Mapped_GCK    9802     14   #1  1179                            NaN   
Sales_EUR   9802.0    NaN  NaN   NaN                  272785.686663   

                            min                  25%                  50%  \
DATE        2018-10-01 00:00:00  2019-08-30 00:00:00  2020-07-26 12:00:00   
Mapped_GCK                  NaN                  NaN                  NaN   
Sales_EUR            -506381.17                  0.0                  0.0   

                            75%                  max           std  
DATE        2021-06-09 00:00:00  2022-04-30 00:00:00           NaN  
Mapped_GCK                  NaN                  NaN           NaN  
Sales_EUR                   0.0          41127988.02  2230789.1619

Based on this, we get the following:

- `Date`: the first recorded sales was in the 1st of October of 2018 and the last one was on the 30th of April of 2022. More or less 3 years and half of reorded sales.
- `Mapped_GCK`: seems that product #1 was selled the most (1179 times).
- `Sales_EUR`: The average sales value is approximately 272,785.69 EUR. The lowest recorded sales value is -506,381.17 EUR. A negative value suggests there were some instances of sales being recorded as refunds or returns leading to negative sales figures. 75% of the entries have sales below 0 EUR. The highest sales recorded is 41,127,988.02 EUR, which is significantly higher than the average. This is likely an outlier. The standard deviation is 2,230,789.16 EUR, which is quite high, indicating a large variability in the sales values.



<div class="alert-danger">

# To Do

- fazer o mesmo que fiz mas para o market data
- analisar cada variável para ambos os datasets e fazer visualizações decentes